File: lab6-openmp/lab6-openmp_cxx/omp_bug5.cpp
/* FILE: omp_bug5.cpp
   DESCRIPTION:
      This program fails due to incorrectly scoped variables.  
      Compare to omp_orphan.c.
   AUTHOR: Blaise Barney  6/05
   UPDATED: Daniel R. Reynolds (updated to C++), 1/13/2013 */

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#define VECLEN 100

// global variables
float a[VECLEN], b[VECLEN];

// external function
float dotprod() {

  // local variables
  int i, tid;
  float sum;

  tid = omp_get_thread_num();
# pragma omp for reduction(+:sum)
  for (i=0; i<VECLEN; i++) {
    sum = sum + (a[i]*b[i]);
    printf("  tid = %i,  i = %i\n", tid, i);
  }
  return sum;
}


// main routine
int main (int argc, char *argv[]) {

  // local variables
  int i;
  float sum;

  // initialize values
  for (i=0; i<VECLEN; i++)   a[i] = b[i] = 1.0 * i;
  sum = 0.0;

  // begin parallel region, with orphaned call to dotprod()
# pragma omp parallel shared(sum)
  sum = dotprod();

  // output result
  printf("Sum = %g\n", sum);

  return 0;
} // end main

File: lab6-openmp/lab6-openmp_f90/omp_bug5.f90
! -*- Mode: Fortran90; -*-
!=================================================================

program Bug6
  !-----------------------------------------------------------------
  ! FILE: omp_bug6.f90
  ! DESCRIPTION:
  !   This program fails due to incorrectly scoped variables.  
  !   Compare to omp_orphan.c.
  ! AUTHOR: Blaise Barney  6/05
  ! Upgraded to F90, free-source form: Daniel R. Reynolds, 2/2009
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: i
  real*8 :: sum
  integer, parameter :: N = 100
  real*8 :: A(N), B(N)

  !======= Internals ============

  do i=1, N
     A(i) = 1.d0 * i
     B(i) = A(i)
  enddo
  sum = 0.d0

  !$omp parallel shared(sum)
  call DotProd(A, B)
  !$omp end parallel

  write(*,*) "Sum = ", sum

end program Bug6
!=================================================================



subroutine DotProd(A, B)
  !-----------------------------------------------------------------
  ! DOTPROD: computes the dot-product of the two vectors A and B.
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: i, TID
  integer, external :: omp_get_thread_num
  integer, parameter :: N = 100
  real*8 :: A(N), B(N), sum

  !======= Internals ============

  TID = omp_get_thread_num()
  !$omp do reduction(+:sum)
  do i=1, N
     sum = sum + (A(i)*B(i))
     print *, '  TID= ',TID,'i= ',i
  enddo

  return

end subroutine DotProd
!=================================================================
